compiler/rustc_monomorphize/src/collector.rs

```diff
@@ -1359,10 +1366,13 @@ fn create_mono_items_for_default_impls<'tcx>(
             continue;
         }
 
-        let substs = InternalSubsts::for_item(tcx, method.def_id, |param, _| match param.kind {
+        let substs = trait_ref.substs.extend_to(tcx, method.def_id, |param, _| match param.kind {
             GenericParamDefKind::Lifetime => tcx.lifetimes.re_erased.into(),
             GenericParamDefKind::Type { .. } | GenericParamDefKind::Const { .. } => {
-                trait_ref.substs[param.index as usize]
+                unreachable!(
+                    "`own_requires_monomorphization` check means that \
+                    we should have no fn type/const params"
+                )
             }
         });
```

Comment:
Maybe create a closure for both uses of this logic?
r=me with a comment explaining the use of erased regions.